tests/erc721/erc721_test.go

```diff
@@ -162,6 +171,289 @@ func TestOpenSeaProxyApproval(t *testing.T) {
+		t.Errorf("After %T deployment and single ownership transfer; Transfer events diff (-want +got):\n%s", filter, diff)
+	}
+}
+
+func TestOpenSeaProxyPreApproval(t *testing.T) {
+	sim, _, _ := deploy(t)
+
+	tests := []struct {
+		name            string
+		from            int
+		to              int
+		hasProxy        bool
+		wantPreApproved bool
+		wantEvents      []*ERC721ApprovalForAll
+	}{
+		{
+			name:            "mint without proxy",
+			from:            -1,
+			to:              tokenReceiver,
+			hasProxy:        false,
+			wantPreApproved: false,
+			wantEvents:      nil,
+		},
+		{
+			name:            "mint with proxy",
+			from:            -1,
+			to:              tokenReceiver,
+			hasProxy:        true,
+			wantPreApproved: true,
+			wantEvents: []*ERC721ApprovalForAll{
+				{
+					Owner:    sim.Addr(tokenReceiver),
+					Operator: sim.Addr(proxy),
+					Approved: true,
+				},
+			},
+		},
+		{
+			name:            "transfer without proxy",
+			from:            tokenOwner,
+			to:              tokenReceiver,
+			hasProxy:        false,
+			wantPreApproved: false,
+			wantEvents:      nil,
+		},
+		{
+			name:            "transfer with proxy",
+			from:            tokenOwner,
+			to:              tokenReceiver,
+			hasProxy:        true,
+			wantPreApproved: true,
+			wantEvents: []*ERC721ApprovalForAll{
+				{
+					Owner:    sim.Addr(tokenReceiver),
+					Operator: sim.Addr(proxy),
+					Approved: true,
+				},
+			},
+		},
+		{
+			name:            "burn without proxy",
+			from:            tokenOwner,
+			to:              -1,
+			hasProxy:        false,
+			wantPreApproved: false,
+			wantEvents:      nil,
+		},
+		{
+			name:            "burn with proxy",
+			from:            tokenOwner,
+			to:              -1,
+			hasProxy:        false,
+			wantPreApproved: false,
+			wantEvents:      nil,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			sim, nft, filter := deploy(t)
+			start := sim.BlockNumber().Uint64()
+
+			toAddr := common.HexToAddress("0x0")
+			if tt.to >= 0 {
+				toAddr = sim.Addr(tt.to)
+			}
+
+			if tt.hasProxy {
+				openseatest.SetProxyTB(t, sim, toAddr, sim.Addr(proxy))
+			}
+
+			if tt.from == -1 {
+				sim.Must(t, "Mint(%d)", notExists)(nft.Mint(sim.Acc(tt.to), big.NewInt(notExists)))
+			} else if tt.to == -1 {
+				sim.Must(t, "Burn(%d)", exists)(nft.Burn(sim.Acc(tt.from), big.NewInt(exists)))
+			} else {
+				sim.Must(t, "TransferFrom(%d)", exists)(nft.TransferFrom(sim.Acc(tt.from), sim.Addr(tt.from), toAddr, big.NewInt(exists)))
+			}
+
+			testApprovalForAllEvents(t, filter, start, tt.wantEvents)
+
+			{
+				got, err := nft.IsApprovedForAll(nil, toAddr, sim.Addr(proxy))
+				if err != nil || got != tt.wantPreApproved {
+					t.Errorf("IsApprovedForAll([owner], [proxy]) got %t, err = %v; want %t, nil err", got, err, tt.wantPreApproved)
+				}
+			}
+		})
+	}
+}
+
+func TestOpenSeaProxyPreApprovalOptInOut(t *testing.T) {
+	sim, _, _ := deploy(t)
```

Comment:
This is actually necessary because we need to call `sim.Addr` to fill the expected events in the table.
I also didn't like adding it here but I couldn't find another solution within reasonable effort. Any idea how this could be avoided?

Reply:
Makes sense. We can just create a new SimulatedBackend and use the addresses, instead of deploying the entire contract. I've noted it down and will submit changes after I merge this.